Resiliency: Add retry logic to attempt to clear out stale hostip

derailed opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

Resiliency: Add retry logic to attempt to clear out stale hostip

NOTE: Depends on PR: https://github.com/cilium/cilium/pull/27614

When the daemon initializes it attempts to clean out stale host ips. 
This process only occurs one time irregardless of errors occurring during that reconciliation.

- Add retry logic on the daemon to attempt to clear out stale host ips should errors occurred during that process.
    
> NOTE: Not super sure on the best way to handle this use case??
This will potentially put the brakes on while constructing the daemon.
This ctor code is currently very procedural (not to mention a mile long!) 
as each step depends on the resolution of the previous one?
    
Fixes: #issue-number

Signed-off-by: Fernand Galiana <fernand.galiana@gmail.com>

bimmlerd commented on daemon/cmd/daemon.go:
```diff
@@ -19,6 +19,7 @@ import (
 	"github.com/sirupsen/logrus"
 	"github.com/vishvananda/netlink"
 	"golang.org/x/sync/semaphore"
+	"k8s.io/apimachinery/pkg/util/wait"
```
I think we have `pkg/backoff` in cilium for exponential backoffs, did you consider using that instead of k8s?

derailed replied:
@bimmlerd  Thanks for your input David! Best I can tell `pkg/backoff` expects you to manage the retry loop, thus opted for using k8s retry built-in instead which felt more appropriate for this use case.

jibi replied:
I would have also suggested to use `pkg/backoff`, mostly for consistency with the rest of the codebase, although I do agree that `wait` seems more handy to deal with retries :+1:

bimmlerd commented on daemon/cmd/daemon.go:
```diff
@@ -1065,13 +1059,30 @@ func newDaemon(ctx context.Context, cleaner *daemonCleanup, params *daemonParams
 	}
 
+	bo := wait.Backoff{
+		Duration: 500 * time.Millisecond,
```
I think that's a long time to wait before retrying given this is on the critical path for daemon startup, I'd shorten this.

derailed replied:
Good point! I am not versed with netlink hence not sure what would be a reasonable delay to retry the service? Lowering it down to 100ms. Please pipe in if you think otherwise. Thank you!

bimmlerd reviewed (changes requested):
Given we've just logged a warning before, I think we should delay startup by as much. generally lgtm, though I have no idea whether these errors should be considered retryable

jibi commented on daemon/cmd/daemon.go:
```diff
@@ -1065,13 +1059,30 @@ func newDaemon(ctx context.Context, cleaner *daemonCleanup, params *daemonParams
-	}
+	ipv6, router := false, router4FromK8s
 	if option.Config.EnableIPv6 {
-		d.restoreCiliumHostIPs(true, router6FromK8s)
```
I think this drifts from the original original behavior: in case both ipv4 and ipv6 are enabled, only v6 addresses will be restored, while before both v4 and v6 would have been restored

derailed replied:
Rats! Great catch. Thank you Gilberto!

ti-mo commented on daemon/cmd/daemon.go:
```diff
@@ -343,25 +345,21 @@ func (d *Daemon) restoreCiliumHostIPs(ipv6 bool, fromK8s net.IP) {
-	} else if err != nil {
-		return err
+	if err != nil {
+		if errors.As(err, &netlink.LinkNotFoundError{}) && restoredIP == nil {
```
Nit: `if errors.As() ... { return nil }` can be lifted out of the `if err != nil {}` block, it already does its own nil checking. If the `As` doesn't match, execution will fall through to the regular `if err != nil {}`, where the retry is issued.

Less indentation reads a bit easier.

derailed replied:
Indeed. Great catch!

ti-mo commented on daemon/cmd/daemon.go:
```diff
@@ -343,25 +345,21 @@ func (d *Daemon) restoreCiliumHostIPs(ipv6 bool, fromK8s net.IP) {
+			// there's no restored IP anyway.
+			return nil
+		}
+		return resiliency.NewRetryableErr(err)
```
Could this be named `resiliency.Retry()` instead? It makes sense to use the imperative 'retry' here, since we want to signal an underlying layer to retry the operation. It also saves some typing and reading. :) I'd name the wrapper type `type Retryable` for brevity as well. Perhaps at some point it will hold more than just an error.

derailed replied:
Good point Timo! I am open to a rename but I think for the time being the naming make sense. 
`Retry()` feels a bit more obscure since we are not performing an actual retry here but wrapping the error to potentially trap it further up the chain. 
In the future, should we opt that additional context around the condition would be deemed useful we could rename accordingly?

ti-mo replied:
Sure, then what about popping the `New` and/or `Err` parts off of the function name? We'd end up with `err = errors.Join(err, resiliency.Retryable(fmt.Errorf("failed to remove IP %s: %w", a.IP, e)))`. That pairs up nicely with `if resiliency.IsRetryable(errs) { ... }`. (Retryable() in the callee caught by IsRetryable in the caller)

And since we're exposing `IsRetryable()` as the canonical way of checking retryable errors, we should also unexport `type RetryableErr`.

> In the future, should we opt that additional context around the condition would be deemed useful we could rename accordingly?

By that time, there will be so many users that this turns into a huge change, meaning increased pain during backports. Let's just get this right from the beginning. :slightly_smiling_face:

derailed replied:
Ok I'll make it so. Tx Timo!

ti-mo commented on daemon/cmd/daemon.go:
```diff
@@ -1065,13 +1059,29 @@ func newDaemon(ctx context.Context, cleaner *daemonCleanup, params *daemonParams
+		Jitter:   0.1,
+		Steps:    3,
+	}
+	err = wait.ExponentialBackoffWithContext(ctx, bo, func(ctx context.Context) (done bool, err error) {
```
Will this retry forever? At which point do we give up and crash the agent? (or drop the operation?)

derailed replied:
This will retry at most 3 times before we give up.

christarazi commented on daemon/cmd/daemon.go:
```diff
@@ -343,12 +345,7 @@ func (d *Daemon) restoreCiliumHostIPs(ipv6 bool, fromK8s net.IP) {
 	restoredIP := node.RestoreHostIPs(ipv6, fromK8s, fromFS, cidrs)
-	if err := removeOldRouterState(ipv6, restoredIP); err != nil {
-		log.WithError(err).Warnf(
-			"Failed to remove old router IPs (restored IP: %s) from cilium_host. Manual intervention is required to remove all other old IPs.",
-			restoredIP,
```
I think it can still be nice to know what the restored IP is in the log msg so that the user can know which old IPs to remove

derailed replied:
No loss here, as we moved the log message further up the stack. 
Hence failed deletions will still log the respective host IPs that could not be deleted.

christarazi commented on daemon/cmd/daemon.go:
```diff
@@ -1065,13 +1059,29 @@ func newDaemon(ctx context.Context, cleaner *daemonCleanup, params *daemonParams
+		return true, errs
+	})
+	if err != nil {
+		log.WithError(err).Error("Restore host ips failed. Manual intervention is required to remove all other old IPs.")
```
```suggestion
		log.WithError(err).Error("Restore of the cilium_host IP failed. Manual intervention is required to remove all other old IPs.")
```

Just to not differ too much from the original log msg.

derailed replied:
Good catch! Tx Chris!!

christarazi commented on daemon/cmd/daemon.go:
```diff
@@ -1065,13 +1059,29 @@ func newDaemon(ctx context.Context, cleaner *daemonCleanup, params *daemonParams
+		Jitter:   0.1,
+		Steps:    3,
+	}
+	err = wait.ExponentialBackoffWithContext(ctx, bo, func(ctx context.Context) (done bool, err error) {
```
I feel like this boilerplate should ideally be collapsed down into a helper function, so that the daemon code doesn't just increasingly become longer each time we add resiliency to the restore paths.

derailed replied:
Indeed! I'll float an idea here and we can backfill the other prs if the proposal is deemed acceptable...

christarazi reviewed (approved):
Very clean now with the helper!

derailed committed:
Resiliency retry helper

Retrying a given call to improve system resiliency is a common pattern.
This helper leverages exponential retries to provide convenience for a
given call to be retried a finite number of times.

Signed-off-by: Fernand Galiana <fernand.galiana@isovalent.com>

derailed committed:
Resiliency: Add retry logic to attempt to clear out stale hostip

- Add a retry on the daemon to attempt to clear out stale host ips

> NOTE: Not super sure on the best way to handle this use case??
This will potentially put the brakes while constructing the daemon.
This ctor code is currently very procedural (not to mention
a mile long!) as each step depends on the resolution of the previous one?

Signed-off-by: Fernand Galiana <fernand.galiana@isovalent.com>

jibi commented on daemon/cmd/daemon.go:
```diff
@@ -343,12 +344,7 @@ func (d *Daemon) restoreCiliumHostIPs(ipv6 bool, fromK8s net.IP) {
-			restoredIP,
-		)
-	}
+	return removeOldRouterState(ipv6, restoredIP)
```
nit: since all errors returned by `removeOldRouterState` are retryable, can we just decorate the eventual returned error here to make this more explicit (and avoid all the changes to the function)?

derailed replied:
@jibi - not keen on this idea as it should be a `explicit` developer choice here vs assuming all errors return by this call will be deemed retryable. If we were to return a non retryable error later on, devs might not be aware that error will be retried.
